test/compile-test.cc

```diff
@@ -53,24 +53,76 @@ TEST(iterator_test, truncating_back_inserter) {
   EXPECT_EQ(buffer, "42");
 }
 
+TEST(compile_test, compile_fallback) {
+  // FMT_COMPILE should fallback on runtime formatting when `if constexpr` is
+  // not available.
+  EXPECT_EQ("42", fmt::format(FMT_COMPILE("{}"), 42));
+}
+
+#ifdef __cpp_if_constexpr
 struct test_formattable {};
 
 FMT_BEGIN_NAMESPACE
 template <> struct formatter<test_formattable> : formatter<const char*> {
+  basic_format_specs<char> specs;
+  detail::arg_ref<char> width_ref;
+  char word_spec = 'f';
+
+  struct spec_handler {
+    formatter& f;
+    basic_format_parse_context<char>& context;
+
+    void on_error(const char* msg) { FMT_THROW(format_error(msg)); }
+    FMT_CONSTEXPR void on_fill(basic_string_view<char> fill) {
+      f.specs.fill = fill;
+    }
+    FMT_CONSTEXPR void on_align(align_t align) { f.specs.align = align; }
+    FMT_CONSTEXPR void on_width(int width) { f.specs.width = width; }
+    FMT_CONSTEXPR void on_precision(int) {}
+    FMT_CONSTEXPR void end_precision() {}
+    template <typename Id>
+    FMT_CONSTEXPR detail::arg_ref<char> make_arg_ref(Id arg_id) {
+      context.check_arg_id(arg_id);
+      return detail::arg_ref<char>(arg_id);
+    }
+    FMT_CONSTEXPR detail::arg_ref<char> make_arg_ref(
+        basic_string_view<char> arg_id) {
+      context.check_arg_id(arg_id);
+      return detail::arg_ref<char>(arg_id);
+    }
+    FMT_CONSTEXPR detail::arg_ref<char> make_arg_ref(detail::auto_id) {
+      return detail::arg_ref<char>(context.next_arg_id());
+    }
+    template <typename Id> FMT_CONSTEXPR void on_dynamic_width(Id arg_id) {
+      f.width_ref = make_arg_ref(arg_id);
+    }
+    template <typename Id> FMT_CONSTEXPR void on_dynamic_precision(Id) {}
+  };
```

Comment:
This is a bit of an overkill for the test. Let's do some basic parsing instead of relying on internal API.

Reply:
So, commits 1+2 then, dropping the 3

Reply:
Yeah, could you update the PR?

Reply:
Of course, I will do so, but maybe an hour later

Reply:
Done